File: src/c++/700-799/704_binary_search.cpp
/**
 * Solution to 704. Binary Search (https://leetcode.com/problems/binary-search/)
 * Difficulty: EASY
 * Runtime: O(log n)
 * Submission: https://leetcode.com/submissions/detail/597687066/
 * 
 * @author jagwirecode
*/

class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        int size = nums.size();

        // Check size is of at least 1
        if (size < 1)
        {
            return -1;
        }

        int low = 0;
        int high = size - 1;
        int mid;

        // Keep looping until left hand side and right hand side are the same
        while (low <= high)
        {
            mid = (low + high) / 2;

            // Check left side
            if (nums.at(mid) < target)
            {
                low = mid + 1;
            }
            // Check right side
            else if (nums.at(mid) > target)
            {
                high = mid - 1;
            }
            // Element found
            else
            {
                return mid;
            }
        }

        // Element not found, return -1
        return -1;
    }
};

File: src/python/700-799/704_binary_search.py
"""
Solution to 704. Binary Search (https://leetcode.com/problems/binary-search/)
Difficulty:     EASY
Runtime:        O(log n) - Logarithmic time
Space:          O(1) - Constant space
Submission:     https://leetcode.com/submissions/detail/605815910/
"""
__author__      = "jagwirecode"

from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        """
        Performs binary search on nums in search for target

        Args:
            nums (List[int]): List of numbers sorted ascendingly to be searched through
            target [int]: The number being searched for

        Returns:
            [int]: Index of target or -1 if not present
        """
        low = 0;
        high = len(nums) - 1;
        mid = -1

        # Two pointer binary search approach         
        while(low <= high):
            mid = low + ((high - low) // 2)
            
            if(nums[mid] < target):
                low = mid + 1
            elif(nums[mid] > target):
                high = mid - 1
            else:
                return mid
            
        return -1